### PR TITLE
Job state awareness v2

### PR DESCRIPTION
**Pull request by lmars**
Jobs are now listed from the database rather than the cluster, and the scheduler is responsible for updating job state whilst watching hosts.

This currently fails to track one-off jobs which do not have a type because process_type is a required field in the database (so [this](https://github.com/flynn/flynn-controller/blob/job-state-awareness/scheduler/main.go#L243) will return an error). Should we assign those one-off jobs a special type, or make the field not required?

This depends on #18 so includes a commit from that, but I will rebase once that is merged.

### CODE REVIEW

**Comment by titanous** [comment]
Good to rebase.

**Comment by lmars** [comment]
@titanous rebased

**Comment by titanous** [comment]
process_type may be null to account for one-off jobs.

File: jobs.go [hunk 1]
```diff
@@ -16,41 +16,92 @@ import (
...
-		}
-	}
+	r.JSON(200, list)
+}
 
-	r.JSON(200, jobs)
+func putJob(job ct.Job, app *ct.App, repo *JobRepo, r ResponseHelper) {
+	job.AppID = app.ID
+	err := repo.Add(&job)
+	if err != nil {
```
Comment (titanous) [inline jobs.go]:
``` go
if err := repo.Add(&job); err != nil {
```

File: scheduler/main.go [hunk 2]
```diff
@@ -223,13 +224,32 @@ func (c *context) watchHost(id string, events chan<- *JobRemovalEvent) {
...
 	for event := range ch {
-		if event.Event != "error" && event.Event != "stop" {
-			continue
-		}
 		job := c.jobs.Get(id, event.JobID)
 		if job == nil {
 			continue
 		}
+
+		j := &ct.Job{ID: event.JobID, AppID: job.Formation.AppID, ReleaseID: job.Formation.Release.ID, Type: job.Type}
```
Comment (titanous) [inline scheduler/main.go]: Does the `ID` need to have the host id included?

File: jobs.go (hunk 1 above)
Comment (lmars) [inline jobs.go]: updated: 2b11d52

File: scheduler/main.go (hunk 2 above)
Comment (lmars) [inline scheduler/main.go]: The job ID goes through quite a cycle before coming back here as an event, but yes looks like it never gets the host id prefixed. Updated: 99d31a7

**Comment by lmars** [comment]
@titanous process_type is no longer required, and I updated the test to cover one-off jobs: df73bdb

File: scheduler/scheduler_test.go [hunk 3]
```diff
@@ -219,18 +232,51 @@ func (s *S) TestWatchHost(c *C) {
 	cl.SetHostClient(hostID, hc)
 
 	cx := newContext(cc, cl)
-	events := make(chan *JobRemovalEvent)
+	events := make(chan *host.Event, 4)
```
Comment (titanous) [inline scheduler/scheduler_test.go:235]: Is there a reason to buffer this channel?

Comment (lmars) [inline scheduler/scheduler_test.go:235]: Without the buffer I was getting a deadlock where both this and the watchHost goroutine were waiting for a host event

File: scheduler/scheduler_test.go [hunk 4]
```diff
@@ -26,6 +26,7 @@ func newFakeControllerClient(appID string, release *ct.Release, artifact *ct.Art
 		formations: map[formationKey]*ct.Formation{
 			formationKey{appID, release.ID}: {AppID: appID, ReleaseID: release.ID, Processes: processes},
 		},
+		jobs:   map[string]*ct.Job{},
```
Comment (titanous) [inline scheduler/scheduler_test.go]:
``` go
make(map[string]*ct.Job)
```

File: scheduler/scheduler_test.go (hunk 3 above)
Comment (titanous) [inline scheduler/scheduler_test.go:235]: Ah, yep, I see why.

File: scheduler/scheduler_test.go (hunk 4 above)
Comment (lmars) [inline scheduler/scheduler_test.go]: done: f81b9a0

**Comment by titanous** [comment]
Nice work!

**Pull request merged**